### PR TITLE
PHP-109- Added isArchived and setArchived and passed tests

### PR DESCRIPTION
**Pull request by G4MAlexVannorsdall**
## Description
I have updated the ShoppingList class with isArchived and setArchived, then went through the ShoppingList class tests, making sure that they all pass. Also added the test for the setter to make sure if a list is set archived to true it will pass. 

## How to run
Checkout as normal, using the terminal.

## Implementation
Help from you and going through older code.

## Thoughts/Considerations
I decided to make isArchived optional and the default value false.

### CODE REVIEW

File: src/Domain/ShoppingList.php [hunk 1]
```diff
@@ -24,16 +24,23 @@ class ShoppingList
...
      * @param ShoppingItemStack|null $items
+     * @param bool|false $isArchived
      */
-    public function __construct(Slug $slug, string $name, ShoppingItemStack $items = null)
+    public function __construct(Slug $slug, string $name, bool $isArchived = false, ShoppingItemStack $items = null)
     {
         $this->slug = $slug;
         $this->setName($name);
         $this->items = $items ?? new ShoppingItemStack();
+        $this->isArchived = $isArchived;
```
Comment (lindyhopchris) [inline src/Domain/ShoppingList.php:42]: Suggestion: really minor point, we would tend to assign these in the order they are arguments for the constructor. I.e. the `$this->isArchived` assignment should got above the `$this->items` assignment.

File: tests/Unit/Domain/ShoppingListTest.php [hunk 2]
```diff
@@ -15,12 +15,13 @@ public function testConstructWithoutList(): ShoppingList
     {
         $list = new ShoppingList(
             $slug = new Slug('my-list'),
-            'My List',
+            'My List'
```
Comment (lindyhopchris) [inline tests/Unit/Domain/ShoppingListTest.php]: Comment: the trailing comma should not have been removed here. It's good not to change lines that have nothing to do with your changes, plus the trailing comma here is correct.

File: src/Domain/ShoppingList.php [hunk 3]
```diff
@@ -80,6 +87,27 @@ public function getItems(): ShoppingItemStack
...
+        return $this->isArchived;
+    }
+
+    /**
+     * Set whether an item is archived.
+     *
+     * @param bool $isArchived
+     * @return void
+     */
+    public function setArchived(bool $isArchived): void
```
Comment (lindyhopchris) [inline src/Domain/ShoppingList.php]: Required: good job on adding the setter, this is correct! Just one change required. The `setName()` method does `return $this`. We'd expect setters on the class to be consistent, so you should also do the same for this method.

File: tests/Unit/Domain/ShoppingListTest.php [hunk 4]
```diff
@@ -54,11 +56,23 @@ public function testAddItem(): void
...
         $this->assertSame([$item1, $item2], $list->getItems()->all());
     }
+
+    /**
+     * @param ShoppingList $list
+     * @depends testConstructWithoutList
+     */
+    public function testSetArchived(ShoppingList $list): void
+    {
+        $list->setArchived(true);
```
Comment (lindyhopchris) [inline tests/Unit/Domain/ShoppingListTest.php]: Required: once you've made the change to what the setter returns, this line needs to assert that the list is returned by the setter, i.e.:

```php
$this->assertSame($list, $list->setArchived(true));
```

You'll see the same pattern in the test for `setName()`.

File: src/Domain/ShoppingList.php (hunk 3 above)
Comment (lindyhopchris) [reply src/Domain/ShoppingList.php]: Just so you know, the reason for returning `$this` from setters is it allows calling code to use method "chaining". I.e. if I needed to set both the name and archived state I can do:

```php
$list->setName('My Groceries')->setArchived(true);
```

Whereas if we didn't return `$this` from the setters, that code would have to be written:

```php
$list->setName('My Groceries');
$list->setArchived(true);
```

Which gets very repetitive if there's lots of things to set! Just so you know, a "chaining" pattern is a developer choice - i.e. to implement it or not is down to the developer. However, once the pattern is in use on a class (as it is on this class' `setName()` method) the class should consistently implement the pattern.

**Review by lindyhopchris — CHANGES_REQUESTED** [review]
Good job on this, the getter and setter looks good. I've put some comments and some required changes - which are new things to learn rather than any problems with the code.

File: src/Domain/ShoppingList.php [hunk 5]
```diff
@@ -80,6 +87,26 @@ public function getItems(): ShoppingItemStack
...
+    }
+
+    /**
+     * Set whether an item is archived.
+     *
+     * @return self
+     */
+    public function setArchived(): self
+    {
+       return $this;
```
Comment (lindyhopchris) [inline src/Domain/ShoppingList.php]: Required: this method no longer works.

File: tests/Unit/Domain/ShoppingListTest.php [hunk 6]
```diff
@@ -54,11 +56,23 @@ public function testAddItem(): void
...
         $this->assertSame([$item1, $item2], $list->getItems()->all());
     }
+
+    /**
+     * @param ShoppingList $list
+     * @depends testConstructWithoutList
+     */
+    public function testSetArchived(ShoppingList $list): void
+    {
+        $this->assertSame($list, $list->setArchived(false));
```
Comment (lindyhopchris) [inline tests/Unit/Domain/ShoppingListTest.php]: Required: the test here is not set up correctly, which is why you've not picked up on the code in `setArchived()` no longer working. You already know that a list starts off with archived as `false` - so setting it to `false` doesn't change anything.

Given: a shopping list that is not archived...
When: I set the archived flag to `true`
Then: it is archived (indicated by `isArchived()` returning `true`).

File: tests/Unit/Domain/ShoppingListTest.php [hunk 7]
```diff
@@ -54,11 +56,23 @@ public function testAddItem(): void
...
+        $this->assertSame(false, $list->isArchived());
```
Comment (lindyhopchris) [inline tests/Unit/Domain/ShoppingListTest.php:74]: Suggestion: PHPUnit has two helper assertions for true and false - `assertTrue($actual)` and `assertFalse($actual)`.

**Review by lindyhopchris — CHANGES_REQUESTED** [review]
Unfortunately your changes have broken the `setArchived()` method, and this wasn't picked up because you've modified the set archived test so that it is no longer valid. My comments should point you in the right direction!

File: src/Domain/ShoppingList.php [hunk 8]
```diff
@@ -99,12 +99,15 @@ public function isArchived(): bool
...
      * Set whether an item is archived.
-     *
+     * @param bool $isArchived
      * @return self
      */
-    public function setArchived(): self
+    public function setArchived(bool $isArchived): self
     {
-       return $this;
+        if (!empty($isArchived)) {
```
Comment (lindyhopchris) [inline src/Domain/ShoppingList.php]: Question: not sure why you've put this in? Should we conditionally be setting the value, or should it just be set to the value that is given to the function (which is already type-hinted as `bool`)?

File: tests/Unit/Domain/ShoppingListTest.php [hunk 9]
```diff
@@ -71,8 +71,8 @@ public function testAddItem(): void
      */
     public function testSetArchived(ShoppingList $list): void
     {
-        $this->assertSame($list, $list->setArchived(false));
-        $this->assertSame(false, $list->isArchived());
+        $this->assertSame($list, $list->setArchived(true));
+        $this->assertTrue(true);
```
Comment (lindyhopchris) [inline tests/Unit/Domain/ShoppingListTest.php]: Required: here you are asserting that true is true, which is always going to pass. You should be asserting that the list is archived.

**Review by lindyhopchris — CHANGES_REQUESTED** [review]
I think we should chat about this on Tuesday morning, because we're circling around the solution rather than reaching it. I'm not entirely sure why so it might be better to discuss rather than going back and forth on the PR? I'll drop you an invite for Tuesday morning to discuss.

**Comment by G4MAlexVannorsdall** [comment]
Pushed the changes to the code we worked on this morning.

**Review by lindyhopchris — APPROVED** [review]

**Pull request merged**